README.md
<!--
SPDX-FileCopyrightText: 2022 Magenta ApS <https://magenta.dk>
SPDX-License-Identifier: MPL-2.0
-->

# Employee Engagement Updater

This repository contains an OS2mo AMQP Trigger that updates employee engagements based on "organisation unit relations."

The intended usage is:
* Organisation units can be related to each other in MO, using the concept of `related_units`.
* This allows specifying a relationship between different organisation hierarchies (e.g. administrative and payroll hierarchies.)
* Engagements created in the payroll hierarchy can then be moved automatically to the corresponding location in the administrative hierarchy.

Program functionality:
* If an engagement is created in the payroll hierarchy, this trigger can detect it, and update the engagement so it belongs to the corresponding organisation unit in the administrative hierarchy instead.
* To mark the origin of the original engagement, an association is created in the original organisation unit.  

## Usage

* Adjust the `AMQP__URL` variable to point to the currently running message broker (RabbitMQ instance) of your OS2mo installation, either:
  * directly in `docker-compose.yml` or
  * by creating a `docker-compose.override.yaml` file.
* Configure the required `ASSOCIATION_TYPE` variable to contain the user key of the desired association type. 
This used when the program creates associations marking the origin of the original engagement.

Now start the container using `docker-compose`:
```
docker-compose up -d
```

You should see the following:
```
2022-10-04 08:08.36 [info     ] Starting metrics server
2022-10-04 08:08.36 [info     ] Setting up clients
2022-10-04 08:08.36 [info     ] Setting up AMQP system
2022-10-04 08:08.36 [info     ] Register called                function=on_amqp_message routing_key=employee.engagement.*
2022-10-04 08:08.36 [info     ] Starting AMQP system
2022-10-04 08:08.36 [info     ] Starting AMQP system
2022-10-04 08:08.36 [info     ] Establishing AMQP connection   host=msg_broker path=/ port=5672 scheme=amqp user=guest
2022-10-04 08:08.36 [info     ] Creating AMQP channel
2022-10-04 08:08.36 [info     ] Attaching AMQP exchange to channel exchange=os2mo
2022-10-04 08:08.36 [info     ] Declaring unique message queue function=on_amqp_message queue_name=os2mo-amqp-trigger-employee-engagement-updater_on_amqp_message
2022-10-04 08:08.36 [info     ] Starting message listener      function=on_amqp_message
2022-10-04 08:08.36 [info     ] Binding routing keys           function=on_amqp_message
2022-10-04 08:08.36 [info     ] Binding routing-key            function=on_amqp_message routing_key=employee.engagement.*
```
Creating an employee engagement in a relevant organisation unit will then log the following:
```
2022-10-04 08:10.48 [debug    ] Received message               function=on_amqp_message message_id=4471787f0fa54055982ffeed1700833c routing_key=employee.engagement.create
2022-10-04 08:10.48 [debug    ] Message received               object_type=engagement payload=PayloadType(uuid=UUID('fceb334b-7fff-40c6-bc30-a10b665af3d2'), object_uuid=UUID('93aae37e-ad69-4fa8-ad9b-69c2a0fd2d76'), time=datetime.datetime(2022, 10, 4, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))) request_type=create service_type=employee
2022-10-04 08:10.49 [debug    ] Found related unit             other_unit=6453b4d3-ced6-4f0c-9720-0353081c94be this_unit=bf6d084e-f211-4322-95aa-40f30f0e6d3b
2022-10-04 08:10.49 [info     ] Created association            engagement_uuid=UUID('93aae37e-ad69-4fa8-ad9b-69c2a0fd2d76') response=dcf8abaf-4cee-40ca-aad1-dad9de3a209f
2022-10-04 08:10.49 [info     ] Updated engagement             engagement_uuid=UUID('93aae37e-ad69-4fa8-ad9b-69c2a0fd2d76') response=93aae37e-ad69-4fa8-ad9b-69c2a0fd2d76
```
And also log:
```
2022-10-04 08:10.49 [debug    ] Received message               function=on_amqp_message message_id=9d96374f78d14972bf370ddf0b1f12ad routing_key=employee.engagement.edit
2022-10-04 08:10.49 [debug    ] Message received               object_type=engagement payload=PayloadType(uuid=UUID('fceb334b-7fff-40c6-bc30-a10b665af3d2'), object_uuid=UUID('93aae37e-ad69-4fa8-ad9b-69c2a0fd2d76'), time=datetime.datetime(2022, 10, 4, 0, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200)))) request_type=edit service_type=employee
2022-10-04 08:10.50 [debug    ] Found related unit             other_unit=bf6d084e-f211-4322-95aa-40f30f0e6d3b this_unit=6453b4d3-ced6-4f0c-9720-0353081c94be
2022-10-04 08:10.50 [info     ] Found association in other unit, doing nothing engagement_uuid=UUID('93aae37e-ad69-4fa8-ad9b-69c2a0fd2d76')
```
As the first request to edit the engagement triggers a subsequent `employee.engagement.edit` message from MO, relating to the "other" organisation unit in a pair of related organisation units.
The trigger logic detects this scenario and logs `Found association in other unit, doing nothing`.
Otherwise an infinite loop of cascading edits would be triggered. 

## Development

### Prerequisites

- [Poetry](https://github.com/python-poetry/poetry)

### Getting Started

1. Clone the repository:
```
git clone git@git.magenta.dk:rammearkitektur/os2mo-triggers/os2mo-amqp-trigger-employee-engagement-updater.git
```

2. Install all dependencies:
```
poetry install
```

3. Set up pre-commit:
```
poetry run pre-commit install
```

### Running the tests

You use `poetry` and `pytest` to run the tests:

`poetry run pytest`

You can also run specific files

`poetry run pytest tests/<test_folder>/<test_file.py>`

and even use filtering with `-k`

`poetry run pytest -k "Manager"`

You can use the flags `-vx` where `v` prints the test & `x` makes the test stop if any tests fails (Verbose, X-fail)

#### Running the integration tests

To run the integration tests, an AMQP instance must be available.

If an instance is already available, it can be used by configuring the `AMQP__URL` environment variable. 
Alternatively a RabbitMQ can be started in docker, using:
```
docker run -d -p 5672:5672 -p 15672:15672 rabbitmq:3-management
```

## Versioning

This project uses [Semantic Versioning](https://semver.org/) with the following strategy:
- MAJOR: Incompatible changes to existing data models
- MINOR: Backwards compatible updates to existing data models OR new models added
- PATCH: Backwards compatible bug fixes

## Authors

Magenta ApS <https://magenta.dk>

## License

This project uses: [MPL-2.0](MPL-2.0.txt)

This project uses [REUSE](https://reuse.software) for licensing.
All licenses can be found in the [LICENSES folder](LICENSES/) of the project.


CHANGELOG.md
<!--
SPDX-FileCopyrightText: 2022 Magenta ApS <https://magenta.dk>
SPDX-License-Identifier: MPL-2.0
-->

CHANGELOG
=========


tests/__init__.py
# SPDX-FileCopyrightText: 2022 Magenta ApS
# SPDX-License-Identifier: MPL-2.0
"""The test package"""
ASSOCIATION_TYPE_USER_KEY = "AssociationTypeUserKey"


tests/conftest.py
# SPDX-FileCopyrightText: 2022 Magenta ApS
# SPDX-License-Identifier: MPL-2.0
"""This module contains pytest specific code, fixtures and helpers."""


tests/test_main.py
# SPDX-FileCopyrightText: 2022 Magenta ApS
# SPDX-License-Identifier: MPL-2.0
# pylint: disable=redefined-outer-name
# pylint: disable=too-many-arguments
"Test `engagement_updater.main`"
import asyncio
from datetime import datetime
from time import monotonic
from typing import Any
from typing import AsyncGenerator
from typing import Callable
from typing import cast
from typing import Generator
from typing import Set
from typing import Tuple
from unittest.mock import ANY
from unittest.mock import AsyncMock
from unittest.mock import call
from unittest.mock import MagicMock
from unittest.mock import patch
from uuid import UUID
from uuid import uuid4

import pytest
from asgi_lifespan import LifespanManager
from fastapi import FastAPI
from fastapi.testclient import TestClient
from ramqp.mo import MOAMQPSystem
from ramqp.mo import MORouter
from ramqp.mo.models import ObjectType
from ramqp.mo.models import PayloadType
from ramqp.mo.models import RequestType
from ramqp.mo.models import ServiceType
from starlette.status import HTTP_200_OK
from starlette.status import HTTP_202_ACCEPTED

from engagement_updater.config import get_settings
from engagement_updater.config import Settings
from engagement_updater.main import build_information
from engagement_updater.main import construct_clients
from engagement_updater.main import create_app
from engagement_updater.main import gather_with_concurrency
from engagement_updater.main import update_build_information
from tests import ASSOCIATION_TYPE_USER_KEY


def get_metric_value(metric: Any, labels: Tuple[str]) -> float:
    """Get the value of a given metric with the given label-set.

    Args:
        metric: The metric to query.
        labels: The label-set to query with.

    Returns:
        The metric value.
    """
    # pylint: disable=protected-access
    metric = metric.labels(*labels)._value
    return cast(float, metric.get())


def clear_metric_value(metric: Any) -> None:
    """Get the value of a given metric with the given label-set.

    Args:
        metric: The metric to query.
        labels: The label-set to query with.

    Returns:
        The metric value.
    """
    metric.clear()


def get_metric_labels(metric: Any) -> Set[Tuple[str]]:
    """Get the label-set for a given metric.

    Args:
        metric: The metric to query.

    Returns:
        The label-set.
    """
    # pylint: disable=protected-access
    return set(metric._metrics.keys())


def test_build_information() -> None:
    """Test that build metrics are updated as expected."""
    clear_metric_value(build_information)
    assert build_information._value == {}  # pylint: disable=protected-access
    update_build_information("1.0.0", "cafebabe")
    assert build_information._value == {  # pylint: disable=protected-access
        "version": "1.0.0",
        "hash": "cafebabe",
    }


async def test_gather_with_concurrency() -> None:
    """Test gather with concurrency."""
    start = monotonic()
    await asyncio.gather(
        *[
            asyncio.sleep(0.1),
            asyncio.sleep(0.1),
            asyncio.sleep(0.1),
        ]
    )
    end = monotonic()
    duration = end - start
    assert duration < 0.15

    start = monotonic()
    await gather_with_concurrency(
        3,
        *[
            asyncio.sleep(0.1),
            asyncio.sleep(0.1),
            asyncio.sleep(0.1),
        ],
    )
    end = monotonic()
    duration = end - start
    assert duration < 0.15

    start = monotonic()
    await gather_with_concurrency(
        1,
        *[
            asyncio.sleep(0.1),
            asyncio.sleep(0.1),
            asyncio.sleep(0.1),
        ],
    )
    end = monotonic()
    duration = end - start
    assert duration > 0.3


@pytest.fixture
def fastapi_app_builder() -> Generator[Callable[..., FastAPI], None, None]:
    """Fixture for the FastAPI app builder."""

    def builder(*args: Any, default_args: bool = True, **kwargs: Any) -> FastAPI:
        if default_args:
            kwargs["client_secret"] = "hunter2"
            kwargs["expose_metrics"] = False
            kwargs["association_type"] = ASSOCIATION_TYPE_USER_KEY
        return create_app(*args, **kwargs)

    yield builder


@pytest.fixture
def fastapi_app(
    fastapi_app_builder: Callable[..., FastAPI]
) -> Generator[FastAPI, None, None]:
    """Fixture for the FastAPI app."""
    yield fastapi_app_builder(client_secret="hunter2", expose_metrics=False)


@pytest.fixture
def test_client_builder(
    fastapi_app_builder: Callable[..., FastAPI]
) -> Generator[Callable[..., TestClient], None, None]:
    """Fixture for the FastAPI test client builder."""

    def builder(*args: Any, **kwargs: Any) -> TestClient:
        return TestClient(fastapi_app_builder(*args, **kwargs))

    yield builder


@pytest.fixture
def test_client(
    test_client_builder: Callable[..., TestClient]
) -> Generator[TestClient, None, None]:
    """Fixture for the FastAPI test client."""
    yield test_client_builder(client_secret="hunter2", expose_metrics=False)


async def test_root_endpoint(test_client: TestClient) -> None:
    """Test the root endpoint on our app."""
    response = test_client.get("/")
    assert response.status_code == 200
    assert response.json() == {"name": "engagement_updater"}


async def test_metrics_endpoint(test_client_builder: Callable[..., TestClient]) -> None:
    """Test the metrics endpoint on our app."""
    test_client = test_client_builder(
        default_args=False,
        client_secret="hunter2",
        association_type=ASSOCIATION_TYPE_USER_KEY,
    )
    response = test_client.get("/metrics")
    assert response.status_code == 200


@patch("engagement_updater.main.handle_engagement_update")
async def test_trigger_all_endpoint(
    handle_engagement_update: MagicMock,
    test_client_builder: Callable[..., TestClient],
) -> None:
    """Test the trigger all endpoint on our app."""
    # Arrange: mock `get_bulk_update_payloads` return value
    async def _async_generator(item: Any) -> AsyncGenerator:
        yield item

    # Arrange: context
    gql_client = AsyncMock()
    model_client = AsyncMock()
    settings = MagicMock(spec=Settings)
    context = {
        "gql_client": gql_client,
        "model_client": model_client,
        "settings": settings,
    }

    # Arrange: mock payload returned by `get_bulk_update_payloads`
    payload = PayloadType(
        uuid=uuid4(),
        object_uuid=uuid4(),
        time=datetime.now(),
    )

    # Act
    with patch("engagement_updater.main.construct_context", return_value=context):
        test_client = test_client_builder()
        with patch(
            "engagement_updater.main.get_bulk_update_payloads",
            return_value=_async_generator(payload),
        ) as get_bulk_update_payloads:
            response = test_client.post("/trigger/all")

    # Assert
    assert response.status_code == HTTP_202_ACCEPTED
    assert response.json() == {"status": "Background job triggered"}
    get_bulk_update_payloads.assert_called_once_with(gql_client)
    handle_engagement_update.assert_called_once_with(
        gql_client,
        model_client,
        ANY,
        ANY,
        payload,
    )


@patch("engagement_updater.main.handle_engagement_update")
async def test_trigger_uuid_endpoint(
    handle_engagement_update: MagicMock,
    test_client_builder: Callable[..., TestClient],
) -> None:
    """Test the trigger uuid endpoint on our app."""
    # Arrange: mock `get_single_update_payload` return value
    async def _async_generator(item: Any) -> AsyncGenerator:
        yield item

    # Arrange: context
    gql_client = AsyncMock()
    model_client = AsyncMock()
    settings = MagicMock(spec=Settings)
    context = {
        "gql_client": gql_client,
        "model_client": model_client,
        "settings": settings,
    }

    # Arrange: mock payload returned by `get_single_update_payload`
    engagement_uuid: UUID = uuid4()
    payload = PayloadType(
        uuid=uuid4(),  # employee UUID
        object_uuid=engagement_uuid,
        time=datetime.now(),
    )

    # Act
    with patch("engagement_updater.main.construct_context", return_value=context):
        test_client = test_client_builder()
        with patch(
            "engagement_updater.main.get_single_update_payload",
            return_value=_async_generator(payload),
        ) as get_single_update_payload:
            response = test_client.post(f"/trigger/{str(engagement_uuid)}")

    # Assert
    assert response.status_code == HTTP_200_OK
    assert response.json() == {"status": "OK"}
    get_single_update_payload.assert_called_once_with(
        gql_client,
        engagement_uuid,
    )
    handle_engagement_update.assert_called_once_with(
        gql_client,
        model_client,
        ANY,
        ANY,
        payload,
    )


@patch("engagement_updater.main.MOAMQPSystem")
@patch("engagement_updater.main.MORouter")
async def test_lifespan(
    mo_router: MORouter,
    mo_amqpsystem: MOAMQPSystem,
    fastapi_app: FastAPI,
) -> None:
    """Test that our lifespan events are handled as expected."""
    amqp_system = MagicMock()
    amqp_system.start = AsyncMock()
    amqp_system.stop = AsyncMock()

    mo_amqpsystem.return_value = amqp_system  # type: ignore

    router = MagicMock()
    mo_router.return_value = router  # type: ignore

    assert not amqp_system.mock_calls

    # Fire startup event on entry, and shutdown on exit
    async with LifespanManager(fastapi_app):
        assert len(router.mock_calls) == 2
        # Create register calls
        assert router.mock_calls[0] == call.register(
            ServiceType.EMPLOYEE, ObjectType.ENGAGEMENT, RequestType.WILDCARD
        )
        # Register calls
        assert router.mock_calls[1]

        # Clean mock to only capture shutdown changes
        amqp_system.reset_mock()


async def test_liveness_endpoint(test_client: TestClient) -> None:
    """Test the liveness endpoint on our app."""
    response = test_client.get("/health/live")
    assert response.status_code == 204


@pytest.mark.parametrize(
    "amqp_ok,gql_ok,model_ok,expected",
    [
        (True, True, True, 204),
        (False, True, True, 503),
        (True, False, True, 503),
        (True, True, False, 503),
        (True, False, False, 503),
        (False, True, False, 503),
        (False, False, True, 503),
        (False, False, False, 503),
    ],
)
@patch("engagement_updater.main.construct_context")
async def test_readiness_endpoint(
    construct_context: MagicMock,
    test_client_builder: Callable[..., TestClient],
    amqp_ok: bool,
    gql_ok: bool,
    model_ok: bool,
    expected: int,
) -> None:
    """Test the readiness endpoint handles errors."""
    gql_client = AsyncMock()
    if gql_ok:
        gql_client.execute.return_value = {
            "org": {"uuid": "35304fa6-ff84-4ea4-aac9-a285995ab45b"}
        }
    else:
        gql_client.execute.return_value = {
            "errors": [{"message": "Something went wrong"}]
        }

    model_client_response = MagicMock()
    if model_ok:
        model_client_response.json.return_value = [
            {"uuid": "35304fa6-ff84-4ea4-aac9-a285995ab45b"}
        ]
    else:
        model_client_response.json.return_value = "BOOM"
    model_client = AsyncMock()
    model_client.get.return_value = model_client_response

    amqp_system = MagicMock()
    amqp_system.healthcheck.return_value = amqp_ok

    construct_context.return_value = {
        "gql_client": gql_client,
        "model_client": model_client,
        "amqp_system": amqp_system,
    }
    test_client = test_client_builder()

    response = test_client.get("/health/ready")
    assert response.status_code == expected

    assert len(gql_client.execute.mock_calls) == 1
    assert model_client.mock_calls == [call.get("/service/o/"), call.get().json()]
    assert amqp_system.mock_calls == [call.healthcheck()]


@pytest.mark.parametrize(
    "amqp_ok,gql_ok,model_ok,expected",
    [
        (True, True, True, 204),
        (False, True, True, 503),
        (True, False, True, 503),
        (True, True, False, 503),
        (True, False, False, 503),
        (False, True, False, 503),
        (False, False, True, 503),
        (False, False, False, 503),
    ],
)
@patch("engagement_updater.main.construct_context")
async def test_readiness_endpoint_exception(
    construct_context: MagicMock,
    test_client_builder: Callable[..., TestClient],
    amqp_ok: bool,
    gql_ok: bool,
    model_ok: bool,
    expected: int,
) -> None:
    """Test the readiness endpoint handled exceptions nicely."""
    gql_client = AsyncMock()
    if gql_ok:
        gql_client.execute.return_value = {
            "org": {"uuid": "35304fa6-ff84-4ea4-aac9-a285995ab45b"}
        }
    else:
        gql_client.execute.side_effect = ValueError("BOOM")

    model_client_response = MagicMock()
    if model_ok:
        model_client_response.json.return_value = [
            {"uuid": "35304fa6-ff84-4ea4-aac9-a285995ab45b"}
        ]
    else:
        model_client_response.json.side_effect = ValueError("BOOM")
    model_client = AsyncMock()
    model_client.get.return_value = model_client_response

    amqp_system = MagicMock()
    if amqp_ok:
        amqp_system.healthcheck.return_value = True
    else:
        amqp_system.healthcheck.side_effect = ValueError("BOOM")

    construct_context.return_value = {
        "gql_client": gql_client,
        "model_client": model_client,
        "amqp_system": amqp_system,
    }
    test_client = test_client_builder()

    response = test_client.get("/health/ready")
    assert response.status_code == expected


@patch("engagement_updater.main.PersistentGraphQLClient")
def test_gql_client_created_with_timeout(mock_gql_client: MagicMock) -> None:
    """Test that PersistentGraphQLClient is called with timeout setting"""

    # Arrange
    settings = get_settings(
        client_secret="not used",
        association_type=ASSOCIATION_TYPE_USER_KEY,
        graphql_timeout=15,
    )

    # Act
    construct_clients(settings)

    # Assert
    assert 15 == mock_gql_client.call_args.kwargs["httpx_client_kwargs"]["timeout"]
    assert 15 == mock_gql_client.call_args.kwargs["execute_timeout"]


engagement_updater/config.py
# SPDX-FileCopyrightText: 2022 Magenta ApS
# SPDX-License-Identifier: MPL-2.0
# pylint: disable=too-few-public-methods,missing-class-docstring
"""Settings handling."""
import logging
from enum import Enum
from functools import cache
from typing import Any

import structlog
from pydantic import AnyHttpUrl
from pydantic import BaseSettings
from pydantic import Field
from pydantic import parse_obj_as
from pydantic import SecretStr
from ramqp.config import ConnectionSettings


class LogLevel(Enum):
    """Log levels."""

    NOTSET = logging.NOTSET
    DEBUG = logging.DEBUG
    INFO = logging.INFO
    WARNING = logging.WARNING
    ERROR = logging.ERROR
    CRITICAL = logging.CRITICAL


logger = structlog.get_logger()


class EmployeeEngagementUpdaterConnectionSettings(ConnectionSettings):
    """Connection settings specific to `engagement_updater`"""

    queue_prefix = "os2mo-amqp-trigger-employee-engagement-updater"

    # TODO: Ensure we don't crash MO when running somewhat concurrently
    prefetch_count = 1


class Settings(BaseSettings):
    """Settings for `engagement_updater`.

    Note that AMQP related settings are taken directly by RAMQP:
    * https://git.magenta.dk/rammearkitektur/ramqp/-/blob/master/ramqp/config.py
    """

    amqp: EmployeeEngagementUpdaterConnectionSettings = Field(
        default_factory=EmployeeEngagementUpdaterConnectionSettings
    )

    commit_tag: str = Field("HEAD", description="Git commit tag.")
    commit_sha: str = Field("HEAD", description="Git commit SHA.")

    mo_url: AnyHttpUrl = Field(
        parse_obj_as(AnyHttpUrl, "http://mo-service:5000"),
        description="Base URL for OS2mo.",
    )
    client_id: str = Field(
        "engagement_updater", description="Client ID for OIDC client."
    )
    client_secret: SecretStr = Field(..., description="Client Secret for OIDC client.")
    auth_server: AnyHttpUrl = Field(
        parse_obj_as(AnyHttpUrl, "http://keycloak-service:8080/auth"),
        description="Base URL for OIDC server (Keycloak).",
    )
    auth_realm: str = Field("mo", description="Realm to authenticate against")

    dry_run: bool = Field(
        False, description="Run in dry-run mode, only printing what would have changed."
    )

    log_level: LogLevel = LogLevel.DEBUG

    expose_metrics: bool = Field(True, description="Whether to expose metrics.")

    graphql_timeout: int = 120

    association_type: str = Field(
        description=(
            "User key of the association type to use for new associations created by "
            "this program."
        )
    )

    class Config:
        env_nested_delimiter = "__"  # allows setting e.g. AMQP__QUEUE_PREFIX=foo


@cache
def get_settings(*args: Any, **kwargs: Any) -> Settings:
    """Fetch settings object.

    Args:
        args: overrides
        kwargs: overrides

    Return:
        Cached settings object.
    """
    settings = Settings(*args, **kwargs)
    logger.debug("Settings fetched", settings=settings, args=args, kwargs=kwargs)
    return settings


engagement_updater/main.py
# SPDX-FileCopyrightText: 2022 Magenta ApS
# SPDX-License-Identifier: MPL-2.0
"""Event handling."""
from asyncio import gather
from asyncio import Semaphore
from contextlib import asynccontextmanager
from contextlib import AsyncExitStack
from functools import partial
from typing import Any
from typing import AsyncGenerator
from typing import Awaitable
from typing import Callable
from typing import Coroutine
from typing import Tuple
from typing import TypeVar
from uuid import UUID

import structlog
from fastapi import BackgroundTasks
from fastapi import FastAPI
from fastapi import Query
from fastapi import Response
from gql import gql
from more_itertools import one
from prometheus_client import Info
from prometheus_fastapi_instrumentator import Instrumentator
from raclients.graph.client import PersistentGraphQLClient
from raclients.modelclient.mo import ModelClient
from ramqp.mo import MOAMQPSystem
from ramqp.mo import MORouter
from ramqp.mo.models import MORoutingKey
from ramqp.mo.models import ObjectType
from ramqp.mo.models import PayloadType
from ramqp.mo.models import RequestType
from ramqp.mo.models import ServiceType
from ramqp.utils import sleep_on_error
from starlette.status import HTTP_202_ACCEPTED
from starlette.status import HTTP_204_NO_CONTENT
from starlette.status import HTTP_503_SERVICE_UNAVAILABLE

from .config import get_settings
from .config import Settings
from .handler import get_bulk_update_payloads
from .handler import get_single_update_payload
from .handler import handle_engagement_update
from .handler import ResultType


logger = structlog.get_logger()

T = TypeVar("T")

build_information = Info("build_information", "Build information")


def update_build_information(version: str, build_hash: str) -> None:
    """Update build information.

    Args:
        version: The version to set.
        build_hash: The build hash to set.

    Returns:
        None.
    """
    build_information.info(
        {
            "version": version,
            "hash": build_hash,
        }
    )


async def healthcheck_gql(gql_client: PersistentGraphQLClient) -> bool:
    """Check that our GraphQL connection is healthy.

    Args:
        gql_client: The GraphQL client to check health of.

    Returns:
        Whether the client is healthy or not.
    """
    query = gql(
        """
        query HealthcheckQuery {
            org {
                uuid
            }
        }
        """
    )
    try:
        result = await gql_client.execute(query)
        if result["org"]["uuid"]:
            return True
    except Exception:  # pylint: disable=broad-except
        logger.exception("Exception occured during GraphQL healthcheck")
    return False


async def healthcheck_model_client(model_client: ModelClient) -> bool:
    """Check that our ModelClient connection is healthy.

    Args:
        model_client: The MO Model client to check health of.

    Returns:
        Whether the client is healthy or not.
    """
    try:
        response = await model_client.get("/service/o/")
        result = response.json()
        if one(result)["uuid"]:
            return True
    except Exception:  # pylint: disable=broad-except
        logger.exception("Exception occured during GraphQL healthcheck")
    return False


def construct_clients(
    settings: Settings,
) -> Tuple[PersistentGraphQLClient, ModelClient]:
    """Construct clients froms settings.

    Args:
        settings: Integration settings module.

    Returns:
        Tuple with PersistentGraphQLClient and ModelClient.
    """
    gql_client = PersistentGraphQLClient(
        url=settings.mo_url + "/graphql/v2",
        client_id=settings.client_id,
        client_secret=settings.client_secret.get_secret_value(),
        auth_server=settings.auth_server,
        auth_realm=settings.auth_realm,
        execute_timeout=settings.graphql_timeout,
        httpx_client_kwargs={"timeout": settings.graphql_timeout},
    )
    model_client = ModelClient(
        base_url=settings.mo_url,
        client_id=settings.client_id,
        client_secret=settings.client_secret.get_secret_value(),
        auth_server=settings.auth_server,
        auth_realm=settings.auth_realm,
    )
    return gql_client, model_client


def configure_logging(settings: Settings) -> None:
    """Setup our logging.

    Args:
        settings: Integration settings module.

    Returns:
        None
    """
    structlog.configure(
        wrapper_class=structlog.make_filtering_bound_logger(settings.log_level.value)
    )


async def gather_with_concurrency(parallel: int, *tasks: Awaitable[T]) -> list[T]:
    """Asyncio gather, but with limited concurrency.

    Args:
        parallel: The number of concurrent tasks being executed.
        tasks: List of tasks to execute.

    Returns:
        List of return values from awaiting the tasks.
    """
    semaphore = Semaphore(parallel)

    async def semaphore_task(task: Awaitable[T]) -> T:
        async with semaphore:
            return await task

    return await gather(*map(semaphore_task, tasks))


def construct_context() -> dict[str, Any]:
    """Construct request context."""
    return {}


def create_app(  # pylint: disable=too-many-statements
    *args: Any, **kwargs: Any
) -> FastAPI:
    """FastAPI application factory.

    Starts the metrics server, then listens to AMQP messages forever.

    Returns:
        None
    """
    settings = get_settings(*args, **kwargs)
    configure_logging(settings)

    app = FastAPI()

    update_build_information(
        version=settings.commit_tag, build_hash=settings.commit_sha
    )

    if settings.expose_metrics:
        logger.info("Starting metrics server")
        Instrumentator().instrument(app).expose(app)

    context = construct_context()

    context["settings"] = settings

    # pylint: disable=unused-argument
    @asynccontextmanager
    async def lifespan(app: FastAPI) -> AsyncGenerator:
        async with AsyncExitStack() as stack:
            logger.info("Setting up clients")
            gql_client, model_client = construct_clients(settings)
            gql_client = context["gql_client"] = await stack.enter_async_context(
                gql_client
            )
            model_client = context["model_client"] = await stack.enter_async_context(
                model_client
            )

            logger.info("Setting up AMQP system")

            router = MORouter()
            amqp_system = MOAMQPSystem(settings=settings.amqp, router=router)

            @sleep_on_error
            @router.register(
                ServiceType.EMPLOYEE, ObjectType.ENGAGEMENT, RequestType.WILDCARD
            )
            async def on_amqp_message(
                mo_routing_key: MORoutingKey,
                payload: PayloadType,
                **_: Any,
            ) -> ResultType:
                return await handle_engagement_update(
                    gql_client,
                    model_client,
                    settings,
                    mo_routing_key,
                    payload,
                )

            context["amqp_system"] = amqp_system

            logger.info("Starting AMQP system")
            await stack.enter_async_context(amqp_system)

            # Yield to keep the AMQP system open until the ASGI application is closed.
            # Control will be returned to here when the ASGI application is shut down.
            yield

    app.router.lifespan_context = lifespan

    @app.get("/")
    async def index() -> dict[str, str]:
        return {"name": "engagement_updater"}

    @app.post("/trigger/all", status_code=HTTP_202_ACCEPTED)
    async def bulk_update(background_tasks: BackgroundTasks) -> dict[str, str]:
        """Call `handle_engagement_update` on all engagements"""
        gql_client: PersistentGraphQLClient = context["gql_client"]
        handle = _get_curried_handle_engagement_update(context)
        # Create tasks for `handle(payload)` (which is equivalent to
        # `handle_engagement_update(..., payload)`.)
        async for payload in get_bulk_update_payloads(gql_client):
            background_tasks.add_task(handle, payload)  # type: ignore
        return {"status": "Background job triggered"}

    @app.post("/trigger/{uuid}")
    async def single_update(
        uuid: UUID = Query(..., description="UUID of the engagement to process"),
    ) -> dict[str, str]:
        """Call `handle_engagement_update` on specific engagement"""
        gql_client: PersistentGraphQLClient = context["gql_client"]
        handle = _get_curried_handle_engagement_update(context)
        # Iterate over the expected single payload, and call the curried version of
        # `handle_engagement_update` for that single payload.
        async for payload in get_single_update_payload(gql_client, uuid):
            await handle(payload)  # type: ignore
        return {"status": "OK"}

    @app.get("/health/live", status_code=HTTP_204_NO_CONTENT)
    async def liveness() -> None:
        """Endpoint to be used as a liveness probe for Kubernetes."""
        return None

    @app.get(
        "/health/ready",
        status_code=HTTP_204_NO_CONTENT,
        responses={
            "204": {"description": "Ready"},
            "503": {"description": "Not ready"},
        },
    )
    async def readiness(response: Response) -> Response:
        """Endpoint to be used as a readiness probe for Kubernetes."""
        response.status_code = HTTP_204_NO_CONTENT

        healthchecks = {}
        try:
            # Check AMQP connection
            healthchecks["AMQP"] = context["amqp_system"].healthcheck()
            # Check GraphQL connection (gql_client)
            healthchecks["GraphQL"] = await healthcheck_gql(context["gql_client"])
            # Check Service API connection (model_client)
            healthchecks["Service API"] = await healthcheck_model_client(
                context["model_client"]
            )
        except Exception:  # pylint: disable=broad-except
            logger.exception("Exception occured during readiness probe")
            response.status_code = HTTP_503_SERVICE_UNAVAILABLE

        for name, ready in healthchecks.items():
            if not ready:
                logger.warn(f"{name} is not ready")

        if not all(healthchecks.values()):
            response.status_code = HTTP_503_SERVICE_UNAVAILABLE

        return response

    return app


def _get_curried_handle_engagement_update(
    context: dict,
) -> Callable[
    [PersistentGraphQLClient, ModelClient, Settings, MORoutingKey],
    Coroutine[Any, Any, ResultType],
]:
    gql_client: PersistentGraphQLClient = context["gql_client"]
    model_client: ModelClient = context["model_client"]
    settings: Settings = context["settings"]
    mo_routing_key: MORoutingKey = MORoutingKey.from_tuple(
        (ServiceType.EMPLOYEE, ObjectType.ENGAGEMENT, RequestType.EDIT)
    )
    # Create curried version of `handle_engagement_update` which only requires the
    # final `payload` argument.
    return partial(
        handle_engagement_update,
        gql_client,
        model_client,
        settings,
        mo_routing_key,
    )
